oauthenticator/cilogon.py

```diff
@@ -241,144 +273,105 @@ def _validate_allowed_idps(self, proposal):
-            will overwrite any value set through CILogonOAuthenticator.username_claim
-            for this identity provider.
-
-        This function returns the username claim list that will be used for the current user trying to login
-        based on the idp that they have selected. If no `CILogonOAutnenticator.allowed_idps` is set, then
-        `CILogonOAutnenticator.username_claim` will be used.
+        Overrides OAuthenticator.user_info_to_username that relies on
+        username_claim to instead consider idp specific config in under
+        allowed_idps[user_info["idp"]]["username_derivation"].
+
+        Returns a username based on user_info and configuration in allowed_idps
+        under the associated idp's username_derivation config.
         """
-        username_claims = [self.username_claim]
-        if self.additional_username_claims:
-            username_claims.extend(self.additional_username_claims)
-        if self.allowed_idps:
-            selected_idp = user_info["idp"]
-            if selected_idp in self.allowed_idps.keys():
-                # The username_claim which should be used for this idp
-                return [
-                    self.allowed_idps[selected_idp]["username_derivation"][
-                        "username_claim"
-                    ]
-                ]
-            else:
-                return username_claims
-        return username_claims
-
-    def _get_username_from_claim_list(self, user_info, username_claims):
-        username = None
-        for claim in username_claims:
-            username = user_info.get(claim)
-            if username:
-                break
+        # NOTE: The first time we have received user_info is when
+        #       user_info_to_username is called by OAuthenticator.authenticate,
+        #       so we make a check here that the "idp" claim is received and
+        #       that we allowed_idps is configured to handle that idp.
+        #
+        user_idp = user_info.get("idp")
+        if not user_idp:
+            message = "'idp' claim was not part of the response to the userdata_url"
+            self.log.error(message)
+            raise web.HTTPError(500, message)
+        if not self.allowed_idps.get(user_idp):
+            message = f"Login with identity provider {user_idp} is not pre-configured"
+            self.log.error(message)
+            raise web.HTTPError(500, message)
+
+        unprocessed_username = self._user_info_to_unprocessed_username(user_info)
+        username = self._get_processed_username(unprocessed_username, user_info)
 
         return username
 
-    def user_info_to_username(self, user_info):
-        username_claims = self._get_final_username_claim_list(user_info)
-        username = self._get_username_from_claim_list(user_info, username_claims)
+    def _user_info_to_unprocessed_username(self, user_info):
+        """
+        Returns a username from
```

Comment:
I think this comment was left unfinished by mistake?

Reply:
Nice catch, resolved by 69535f0!